apis/mattermost/v1beta1/mattermost_types.go

```diff
@@ -0,0 +1,263 @@
+	// values regardless if set by Size or manually.
+	// +optional
+	Size string `json:"size,omitempty"`
+
+	// Image defines the Mattermost Docker image.
+	Image string `json:"image,omitempty"`
+	// Version defines the Mattermost Docker image version.
+	Version string `json:"version,omitempty"`
+	// Replicas defines the number of replicas to use for the Mattermost app
+	// servers.
+	Replicas int32 `json:"replicas,omitempty"`
+	// Optional environment variables to set in the Mattermost application pods.
+	// +optional
+	MattermostEnv []v1.EnvVar `json:"mattermostEnv,omitempty"`
+	// MattermostLicenseSecret is the name of the secret containing a license.
+	// +optional
+	MattermostLicenseSecret string `json:"mattermostLicenseSecret,omitempty"`
+	// IngressName defines the name to be used when creating the ingress rules
+	IngressName string `json:"ingressName"`
+	// +optional
+	IngressAnnotations map[string]string `json:"ingressAnnotations,omitempty"`
+	// +optional
+	UseIngressTLS bool `json:"useIngressTLS,omitempty"`
+	// +optional
+	UseServiceLoadBalancer bool `json:"useServiceLoadBalancer,omitempty"`
+	// +optional
+	ServiceAnnotations map[string]string `json:"serviceAnnotations,omitempty"`
+	// +optional
+	ResourceLabels map[string]string `json:"resourceLabels,omitempty"`
+	// Volumes allows for mounting volumes from various sources into the
+	// Mattermost application containers.
+	// +optional
+	Volumes []*v1.Volume `json:"volumes,omitempty"`
+
+	// Advanced contains additional advanced settings for Mattermost.
+	Advanced Advanced `json:"advanced,omitempty"`
+
+	// External Services
+	Database      Database      `json:"database,omitempty"`
+	Filestore     Filestore     `json:"filestore,omitempty"`
+	ElasticSearch ElasticSearch `json:"elasticSearch,omitempty"`
+}
+
+// Advanced defines the advanced configuration for Mattermost. These settings
+// should be correct for most types of deployments.
+type Advanced struct {
+	// Defines the resource requests and limits for the Mattermost app server pods.
+	// +optional
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// If specified, affinity will define the pod's scheduling constraints
+	// +optional
+	Affinity *v1.Affinity `json:"affinity,omitempty"`
+	// Defines the probe to check if the application is up and running.
+	// +optional
+	LivenessProbe v1.Probe `json:"livenessProbe,omitempty"`
+	// Defines the probe to check if the application is ready to accept traffic.
+	// +optional
+	ReadinessProbe v1.Probe `json:"readinessProbe,omitempty"`
+}
+
+// Database defines the database configuration for Mattermost.
+type Database struct {
```

Comment:
I wonder what do you think about splitting the fields to two sub structs `SelfManaged/External` and `OperatorManaged` (probably would need better names). I think this would clean up the struct a little bit but I am not sure about all the implications.
The same applies to filestore.

Reply:
Yes, I had the same idea and forgot to mention it in the original PR comment. I do think there is a good opportunity to do that work with the new spec.

I think it may be slightly easier to make these changes as we design the controller logic as I have a few ideas on how it could work, but feel like it may become clearer as we build out the actual deployment logic. Due to this new resource being clearly marked as not-production-ready, we are safe to iterate as we build. What do you think?

Reply:
Yup, I agree, I think oy would be hard to speculate the final form right now as it may be heavily impacted by the implementation.

Reply:
Since we are doing the v1beta1 do you think is better we drop the support for minio operator and MySQL operator? 

less things to we care, and we know there are lots of breaking changes and it is hard to maintain things that we cannot control